### PR TITLE
Allow linking YouTube channel to account

### PR DESCRIPTION
**Pull request by NuckChorris**
Crossposts YouTube videos onto your profile feed

### CODE REVIEW

**Comment by trmcnvn** [comment]
Can you add the new model as a `model_hint` to `LinkedAccountResource` :heart:

**Comment by NuckChorris** [comment]
Ah, nice catch!

File: app/models/linked_account/youtube_channel.rb [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        errors.add(:token, 'could not be verified') unless client.valid?
+      end
+    end
+
+    after_commit do
+      if share_from_changed?
+        if share_from?
+          raise 'Subscribe failed' unless subscription.subscribe
+        else
+          raise 'Unsubscribe failed' unless subscription.unsubscribe
```
Comment (trmcnvn) [inline app/models/linked_account/youtube_channel.rb:54]: Shouldn't we also unsubscribe on destroy?

**Comment by trmcnvn** [comment]
Looks to be missing the policy files and inclusion of the `addressable` gem.

File: app/services/youtube_service/client.rb [hunk 2]
```diff
@@ -0,0 +1,38 @@
+class YoutubeService
+  class Client
+    API_PREFIX = 'https://www.googleapis.com/youtube/v3'.freeze
+    CHANNEL_URL = Addressable::Template.new(
+      "#{API_PREFIX}/channels.list{?query*}"
+    ).freeze
+    VERIFY_URL = Addressable::Template.new(
+      "#{API_PREFIX}/tokeninfo{?query*}"
+    ).freeze
```
Comment (trmcnvn) [inline app/services/youtube_service/client.rb]: Looks like the URL here is actually `https://www.googleapis.com/oauth2/v3/tokeninfo`

Edit: Example response

```json
{
"azp": "411649055059-gff0hi34kivvut786sv8odho0vedmvl3.apps.googleusercontent.com",
"aud": "411649055059-gff0hi34kivvut786sv8odho0vedmvl3.apps.googleusercontent.com",
"scope": "https://www.googleapis.com/auth/youtube.readonly",
"exp": "1490380334",
"expires_in": "3304",
"access_type": "online"
}
```

File: app/services/youtube_service/client.rb [hunk 3]
```diff
@@ -0,0 +1,38 @@
+class YoutubeService
+  class Client
+    API_PREFIX = 'https://www.googleapis.com/youtube/v3'.freeze
+    CHANNEL_URL = Addressable::Template.new(
+      "#{API_PREFIX}/channels.list{?query*}"
+    ).freeze
```
Comment (trmcnvn) [inline app/services/youtube_service/client.rb]: Url here seems to be `https://www.googleapis.com/youtube/v3/channels` (Ref: https://developers.google.com/youtube/v3/docs/channels/list)

Example Response
```json
{
    "kind": "youtube#channelListResponse",
    "pageInfo": {
        "totalResults": 1,
        "resultsPerPage": 1
    },
    "items": [{
        "kind": "youtube#channel",
        "id": "UCcacpXYQrVx6ywO0QxW5nNQ"
    }]
}
```

**Comment by NuckChorris** [comment]
I think that should address everything?

File: app/services/youtube_service/client.rb [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+      res.success? && JSON.parse(res.body)['aud'] == api_key
+    end
+
+    def channel_id
+      res = Typhoeus.get(CHANNEL_URL.expand(query: {
+        part: 'id',
+        mine: 'true',
+        access_token: token
+      }))
+      JSON.parse(res.body)['id'] if res.success?
```
Comment (trmcnvn) [inline app/services/youtube_service/client.rb]: The response body here is in the following format:

```
{
    "kind": "youtube#channelListResponse",
    "pageInfo": {
        "totalResults": 1,
        "resultsPerPage": 1
    },
    "items": [{
        "kind": "youtube#channel",
        "id": "UCcacpXYQrVx6ywO0QxW5nNQ"
    }]
}
```

File: app/services/youtube_service/client.rb [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    CHANNEL_URL = Addressable::Template.new(
+      "#{API_PREFIX}/youtube/v3/channels{?query*}"
+    ).freeze
+    VERIFY_URL = Addressable::Template.new(
+      "#{API_PREFIX}/oauth2/v3/tokeninfo{?query*}"
+    ).freeze
+
+    attr_reader :token
+
+    def initialize(token, api_key:)
```
Comment (trmcnvn) [inline app/services/youtube_service/client.rb]: Getting error here that `api_key` is always missing.

**Review by trmcnvn — CHANGES_REQUESTED** [review]
:construction:

File: app/services/youtube_service/client.rb (hunk 4 above)
Comment (NuckChorris) [reply app/services/youtube_service/client.rb]: In hindsight, I have no clue why I assumed it wouldn't use the standard response format for this 🤔

**Review by trmcnvn — APPROVED** [review]

**Pull request merged**